### PR TITLE
FFI metrics reporter

### PR DESCRIPTION
**Pull request by leohahn**
This PR adds support for implementing a metrics reporter in a FFI language, in our case it is C#.

Currently, the implementation only logs things on the C# side, and does not implement a prometheus interface or something similar. That would be done on another PR.

### CODE REVIEW

File: pitaya-sharp/NPitaya/src/PitayaCluster.API.cs [hunk 1]
```diff
@@ -286,25 +286,25 @@ public static Task<T> Rpc<T>(Route route, object msg)
...
+
+        // public static Task AddGauge(string name, double value, string[] labels = null)
+        // {
+        //     return _metricsReporter.AddGauge(name, value, labels == null ? new string[] { } : labels);
+        // }
+
+        // public static Task SetGauge(string name, double value, string[] labels = null)
+        // {
+        //     return _metricsReporter.SetGauge(name, value, labels == null ? new string[] { } : labels);
+        // }
```
Comment (victor-carvalho) [inline pitaya-sharp/NPitaya/src/PitayaCluster.API.cs]: Could we delete this commented code?

File: pitaya-sharp/exampleapp/Handlers/TestHandler.cs [hunk 2]
```diff
@@ -8,17 +8,23 @@
 #pragma warning disable 1998
 namespace exampleapp.Handlers
 {
+    // class BackendResponse
+    // {
+    //     public int code;
+    //     public string result;
+    // }
+
```
Comment (victor-carvalho) [inline pitaya-sharp/exampleapp/Handlers/TestHandler.cs]: Same with all the commented code here in this file. Could delete?

File: src/pitaya/src/ffi/metrics.rs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    user_data: *mut c_void,
+    mut opts: crate::metrics::Opts,
+) {
+    // TODO(lhahn): this function allocates a lot of unnecessary memory.
+    // consider improving this in the future.
+
+    let mut variable_labels_ptr: Vec<*mut c_char> = opts
+        .variable_labels
+        .iter_mut()
+        .map(|s| s.as_mut_ptr() as *mut c_char)
```
Comment (victor-carvalho) [inline src/pitaya/src/ffi/metrics.rs]: We need to allocate `CString`s for these strings too 😞

File: src/pitaya/src/ffi/metrics.rs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    os::raw::c_char,
+};
+
+#[repr(C)]
+pub struct PitayaMetricsOpts {
+    pub namespace: *const c_char,
+    pub subsystem: *const c_char,
+    pub name: *const c_char,
+    pub help: *const c_char,
+    pub variable_labels: *mut *mut c_char,
```
Comment (victor-carvalho) [inline src/pitaya/src/ffi/metrics.rs]: In our callbacks we are using `*mut *const c_char` to represent a Vec<String> but here we are using `*mut *mut c_char`. I think it should use `*mut *const c_char` to be consistent with the callbacks.

File: src/pitaya/src/ffi/metrics.rs [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    async fn shutdown(&mut self) -> Result<(), crate::metrics::Error> {
+        // TODO(lhahn): consider implementing shutdown for FFI.
+        Ok(())
+    }
+
+    fn inc_counter(&self, name: &str, labels: &[&str]) -> Result<(), crate::metrics::Error> {
+        let mut labels_ptr: Vec<*const c_char> =
+            labels.iter().map(|s| s.as_ptr() as *const c_char).collect();
```
Comment (victor-carvalho) [inline src/pitaya/src/ffi/metrics.rs]: Need to allocate CString too.

File: src/pitaya/src/ffi/metrics.rs [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    fn observe_hist(
+        &self,
+        name: &str,
+        value: f64,
+        labels: &[&str],
+    ) -> Result<(), crate::metrics::Error> {
+        let mut labels_ptr: Vec<*const c_char> =
+            labels.iter().map(|s| s.as_ptr() as *const c_char).collect();
```
Comment (victor-carvalho) [inline src/pitaya/src/ffi/metrics.rs]: Same here.

File: src/pitaya/src/ffi/metrics.rs [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    fn set_gauge(
+        &self,
+        name: &str,
+        value: f64,
+        labels: &[&str],
+    ) -> Result<(), crate::metrics::Error> {
+        let mut labels_ptr: Vec<*const c_char> =
+            labels.iter().map(|s| s.as_ptr() as *const c_char).collect();
```
Comment (victor-carvalho) [inline src/pitaya/src/ffi/metrics.rs]: Same here.

File: src/pitaya/src/ffi/metrics.rs [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    fn add_gauge(
+        &self,
+        name: &str,
+        value: f64,
+        labels: &[&str],
+    ) -> Result<(), crate::metrics::Error> {
+        let mut labels_ptr: Vec<*const c_char> =
+            labels.iter().map(|s| s.as_ptr() as *const c_char).collect();
```
Comment (victor-carvalho) [inline src/pitaya/src/ffi/metrics.rs]: Same here.

**Review by victor-carvalho — CHANGES_REQUESTED** [review]

File: src/pitaya/src/ffi/metrics.rs (hunk 3 above)
Comment (leohahn) [reply src/pitaya/src/ffi/metrics.rs]: Damn it, true :/. I'll change that.

**Review by victor-carvalho — APPROVED** [review]
LGTM

**Pull request merged**